Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+    /// does not throw an error, if the tool isn't found. Extensions that re-
+    /// quire the tool should emit an error diagnostic if it cannot be found.
+    public func lookupTool(named name: String) throws -> Path {
+        // TODO: Rather than just appending the name, this should instead use
+        // a mapping of tool names to paths (passed in from the context).
+        return execsDir.appending(name)
+    }
+    
+    /// A directory in which the built executables available to the extension
+    /// will be located.
+    /// TODO: This should instead be a mapping of tool names to paths (passed
+    /// in from the context).
+    private let execsDir: Path
+}
+
+/// Constructs commands to run during the build, including full command lines.
+/// All paths should be based on the ones passed to the extension in the target
+/// build context.
+public final class CommandConstructor {
+    
+    fileprivate init() {}
+    
+    /// Creates a command to run during the build. The executable should be a
+    /// path returned by `TargetBuildContext.lookupTool(named:)`, and all the
+    /// paths in both the command line and the input and output lists should be
+    /// based on the paths in the target build context structure.
+    ///
+    /// Note that input and output dependencies are ignored for prebuild and
+    /// postbuild actions, since they always run before and after the build
+    /// respectively.
+    public func createCommand(displayName: String?, executable: Path, arguments: [String], workingDirectory: Path? = nil, environment: [String: String]? = nil, inputPaths: [Path] = [], outputPaths: [Path] = [], derivedSourcePaths: [Path] = []) {
+        output.commands.append(Command(executable: executable, arguments: arguments, workingDirectory: workingDirectory, environment: environment, displayName: displayName, inputPaths: inputPaths, outputPaths: outputPaths, derivedSourcePaths: derivedSourcePaths))
+    }
+    
+    func addGeneratedOutputFile(path: Path) {
+        output.generatedFilePaths.append(path.string)
+    }
+    
+    public func addPrebuildOutputDirectory(path: Path) {
+        output.prebuildOutputDirectories.append(path.string)
+    }
+}
+
+/// A constructed command.
+fileprivate struct Command: Encodable {
+    let executable: Path
+    let arguments: [String]
+    let workingDirectory: Path?
+    let environment: [String: String]?
+    let displayName: String?
+    let inputPaths: [Path]
+    let outputPaths: [Path]
+    let derivedSourcePaths: [Path]
+}
+
+/// Emits errors, warnings, and remarks to be shown as a result of running the
+/// extension. After emitting one or more errors, the extension should return a
+/// non-zero exit code.
+public final class DiagnosticsEmitter {
+    
+    fileprivate init() {}
+    
+    public func fatalError(_ message: String, file: Path? = nil, line: Int? = nil) {
+        emit(error: message, file: file, line: line)
+        exit(1)
```

Comment:
intentional?

Reply:
It needs a better implementation, but I expect that unlike the package description, there will need to be a way to do error checking on things like not being able to find a tool and then good error messages.  It could also be a thrown error though.  This is something I'd like to discuss — we didn't talk that much about what a good error API for this use case should look like.

Reply:
Removed this for now.  We can discuss the best API here as a refinement.